jsonrpc/types.go

```diff
@@ -59,45 +59,50 @@ func (h transactionHash) MarshalText() ([]byte, error) {
 }
 
 func toPendingTransaction(t *types.Transaction) *transaction {
-	return toTransaction(t, nil, nil, nil)
+	return toTransaction(t, nil, nil)
 }
 
 func toTransaction(
 	t *types.Transaction,
-	blockNumber *argUint64,
-	blockHash *types.Hash,
+	header *types.Header,
 	txIndex *int,
 ) *transaction {
 	v, r, s := t.RawSignatureValues()
 	res := &transaction{
-		Nonce:       argUint64(t.Nonce()),
-		Gas:         argUint64(t.Gas()),
-		To:          t.To(),
-		Value:       argBig(*t.Value()),
-		Input:       t.Input(),
-		V:           argBig(*v),
-		R:           argBig(*r),
-		S:           argBig(*s),
-		Hash:        t.Hash(),
-		From:        t.From(),
-		Type:        argUint64(t.Type()),
-		BlockNumber: blockNumber,
-		BlockHash:   blockHash,
+		Nonce: argUint64(t.Nonce()),
+		Gas:   argUint64(t.Gas()),
+		To:    t.To(),
+		Value: argBig(*t.Value()),
+		Input: t.Input(),
+		V:     argBig(*v),
+		R:     argBig(*r),
+		S:     argBig(*s),
+		Hash:  t.Hash(),
+		From:  t.From(),
+		Type:  argUint64(t.Type()),
 	}
 
-	if t.GasPrice() != nil && t.Type() != types.DynamicFeeTxType {
-		gasPrice := argBig(*(t.GasPrice()))
-		res.GasPrice = &gasPrice
+	if t.GasPrice() != nil {
+		res.GasPrice = argBigPtr(t.GasPrice())
 	}
 
-	if t.GasTipCap() != nil && t.Type() == types.DynamicFeeTxType {
-		gasTipCap := argBig(*(t.GasTipCap()))
-		res.GasTipCap = &gasTipCap
+	if header != nil {
+		res.BlockNumber = argUintPtr(header.Number)
+		res.BlockHash = &header.Hash
+
+		if res.GasPrice == nil {
```

Comment:
It can still happen that the `GasPrice` won't be set if `t.GasPrice() == nil`, and `header == nil`.

I suggest something like this:

```go
    baseFee = big.NewInt(0)
    
    if header != nil {
	res.BlockNumber = argUintPtr(header.Number)
	res.BlockHash = &header.Hash
        baseFee = header.BaseFee
   }
   
   if t.GasPrice() != nil {
	res.GasPrice = argBigPtr(t.GasPrice())
   } else {
        res.GasPrice = argBigPtr(t.GetGasPrice(baseFee))
   }

```

Reply:
I see your point, but I wouldn't do that, because in case a `header` is nil (which would be the case when a transaction is not yet mined), we would fake the `baseFee` and set it to 0, which would then set the `gasPrice` to 0 as well and that is not correct.

So IMO, it is better to leave it unpopulated in this use case. Do you agree?

P.s. alternatively we could change the GetGasPrice that in case it receives nil for the baseFee (in case there is no header), it returns nil for the `gasPrice` as well. That way, code could be a bit cleaner, and look similarly to what you have proposed. LMK what you think about it.

Reply:
Base fee is faked. It should not be 0, but undefined or at least read from somewhere, since we do not have the header instance. If we set base fee to 0 then the calculated gas price would be set to 0 as well, and it is not really correct, but instead it should be undefined/nil.

Reply:
Fixed (https://github.com/Ethernal-Tech/blade/pull/284/commits/85d398a5e5ee1667199dfc1fe28439537ec59d0e, https://github.com/Ethernal-Tech/blade/pull/284/commits/3b0b10c8d9779b09b1e9a17af2cab3566ff05bbe).

We have agreed it should work like go-ethereum does:
- in case a transaction is pending (namely not mined yet), it would set `gasPrice` value. For dynamic fee tx, it is equivalent to the `gasFeeCap` value
- otherwise (transaction is mined):
   - dynamic fee tx type: `gasPrice = min(gasFeeCap + baseFee, gasTipCap)`
   - legacy, access list, state tx types: `gasPrice` is expected to be populated, so we are returning that value straight away.